lib/portlayer/metrics/container.go

```diff
@@ -0,0 +1,327 @@
+)
+
+const (
+	// number of samples per collection
+	sampleSize = int32(2)
+	// number of seconds between sample collection
+	sampleInterval = int32(20)
+)
+
+// CPUUsage provides individual CPU metrics
+type CPUUsage struct {
+	// processor id (0,1,2)
+	ID int
+	// MhzUsage is the MhZ consumed by a specific processor
+	MhzUsage int64
+}
+
+// CPUMetrics encapsulates available vm CPU metrics
+type CPUMetrics struct {
+	// CPUs are the individual CPU metrics
+	CPUs []CPUUsage
+	// Usage is the percentage of total vm CPU usage
+	Usage float32
+}
+
+// MemoryMetrics encapsulates available vm memory metrics
+type MemoryMetrics struct {
+	// Consumed memory of vm in bytes
+	Consumed int64
+	// Active memory of vm in bytes
+	Active int64
+	// Provisioned memory of vm in bytes
+	Provisioned int64
+}
+
+// NetworkUsage provides metrics for specific networks
+type NetworkUsage struct {
+	//TBD
+}
+
+// NetworkMetrics encapsulates available vm Network metrics
+type NetworkMetrics struct {
+	//TBD
+}
+
+// DiskUsage provides metrics for specific disks
+type DiskUsage struct {
+	//TBD
+}
+
+// NetworkMetrics encapsulates available vm Storage metrics
+type StorageMetrics struct {
+	//TBD
+}
+
+// VMMetrics encapsulates vm metrics available
+type VMMetrics struct {
+	CPU        CPUMetrics
+	Memory     MemoryMetrics
+	Network    NetworkMetrics
+	Storage    StorageMetrics
+	SampleTime time.Time
+	// interval of collection in seconds
+	Interval int32
+}
+
+// collectorVM is the VM metrics collector
+type collectorVM struct {
+	session *session.Session
+	perfMgr *performance.Manager
+
+	// subscribers to streaming
+	mu   sync.RWMutex
+	subs map[types.ManagedObjectReference]*subscriber
+}
+
+// subscriber is the receiver of the metrics
+type subscriber struct {
+	id        string
+	ref       types.ManagedObjectReference
+	publisher *pubsub.Publisher
+}
+
+func newVMCollector(session *session.Session) *collectorVM {
+	s := make(map[types.ManagedObjectReference]*subscriber)
+	cc := &collectorVM{
+		session: session,
+		subs:    s,
+		perfMgr: performance.NewManager(session.Vim25()),
+	}
+
+	// kick off the subscription sampler
+	go cc.sampler()
+
+	return cc
+}
+
+// sampler will check for subscriptions and sample when needed
+func (cc *collectorVM) sampler() {
+	// TODO: replace this with service that only runs when there are
+	// subscribers
+	d := time.Duration(int64(sampleInterval)) * time.Second
+	for range time.Tick(d) {
+		// if we have no subscribers skip
+		if cc.SubscriberCount() == 0 {
+			continue
+		}
+		// collect metrics for current subscribers
+		cc.collect(cc.perfMgr, cc.Subscribers(), false)
+	}
+}
+
+// Sample returns a single metrics collection
+func (cc *collectorVM) Sample(contain interface{}) (chan interface{}, error) {
+	defer trace.End(trace.Begin(""))
+
+	containerSub, err := containerSubscriber(contain)
+	if err != nil {
+		return nil, err
+	}
+
+	// create a manager for this request
+	mgr := performance.NewManager(cc.session.Vim25())
```

Comment:
This specific case is for a non-streaming request -- instead of waiting for the next instance of the subscription perf Manager to handle the request I create a new one so that the user doesn't have to wait up to 20 seconds to get a reply.  I originally had one performance manager that handled all of the requests -- but because of the 20s interval I opted against it.

Additionally, I was going to create an issue stating that we may eventually need a pool of managers as the documentation states that for best performance you should limit the manager to 50 objects.  So from my perspective I don't think we can get to one.

I think for now this works and we can open appropriate issues for a different direction if needed.